include/kernels/madwf_tensor.cuh

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <color_spinor_field.h>
+#include <color_spinor_field_order.h>
+#include <shared_memory_cache_helper.cuh>
+#include <reduce_helper.h>
+#include <kernels/madwf_transfer.cuh>
+#include <fast_intdiv.h>
+#include <madwf_ml.h>
+
+/**
+  @file This file contains the argument and kernel that forms a tensor outer product in the fifth
+  dimension given two input vectors,
+      T_{st} = conj(x_s * y_t),
+  where s and t are fifth dimension indices. T_{st} is a spin matrix so T has shape Ls*4-by-Ls*4.
+*/
+
+namespace quda
+{
+
+  namespace madwf_ml
+  {
+
+    template <class storage_t, int nSpin_, int nColor_>
+    struct Tensor5DReduceArg : ReduceArg<array<complex<typename mapper<storage_t>::type>, nSpin_ * nSpin_>> {
+
+      static constexpr int nSpin = nSpin_;
+      static constexpr int nColor = nColor_;
+
+      using F = typename colorspinor_mapper<storage_t, nSpin, nColor>::type;
+      using real = typename mapper<storage_t>::type;
+      using reduce_t = array<complex<real>, nSpin * nSpin>;
+      using Vector = ColorSpinor<real, nColor, nSpin>;
+
+      const F x; // xput vector field
+      const F y; // yput vector field
+
+      const int_fastdiv Ls_x; // length of 5th dimension of the x field
```

Comment:
Doesn't need to be an `int_fastdiv`

Reply:
https://github.com/lattice/quda/pull/1061/commits/65dc02461126ad5aee630ce9153055476d15d98b